lib/Sema/TypeCheckPattern.cpp

```diff
@@ -360,10 +360,11 @@ class ResolvePattern : public ASTVisitor<ResolvePattern,
     auto cast = dyn_cast<CoerceExpr>(E->getElement(1));
     if (!cast)
       return nullptr;
-    
+
+    const auto tyLoc = TypeLoc(cast->getCastTypeExpr()->getTypeRepr());
     Pattern *subPattern = getSubExprPattern(E->getElement(0));
-    return new (Context) IsPattern(cast->getLoc(), cast->getCastTypeLoc(),
-                                   subPattern, CheckedCastKind::Unresolved);
+    return new (Context) IsPattern(cast->getLoc(), tyLoc, subPattern,
```

Comment:
Until `IsPattern` is disabused of its TypeLoc, you also have to make sure to fill in the instance type from the cast type here.